src/Compilers/CSharp/Test/Syntax/Parsing/FuzzTesting.cs

```diff
@@ -81,18 +81,25 @@ public void c024928()
 
         [WorkItem(2771, "https://github.com/dotnet/roslyn/issues/2771")]
         [Fact]
-        public void TestBinary()
+        public unsafe void TestBinary()
         {
             // use a fixed seed so the test is reproducible
             var random = new System.Random(12345);
-            const int n = 40 * 1000 * 1000; // 40 million "character"s
-            var builder = new StringBuilder(n + 10);
-            for (int i = 0; i < n; i++)
+            // 40 million "character"s
+            const int n = 40 * 1000 * 1000; 
+            var chars = new char[n];
+            for (int i = 0; i < n - 1; i++)
             {
-                builder.Append((char)random.Next(0xffff));
+                chars[i] = (char) random.Next(char.MaxValue);
+            }
+            chars[n - 1] = (char) 0;
+            unsafe
+            {
+                fixed (char* bytePtr = chars)
+                {
+                    var tree = CSharpSyntaxTree.ParseText(new string(bytePtr));
+                }
```

Comment:
What you can do instead, is just create a string of the size you want, and then just wrap it in a fixed block and then directly modify the individual chars (evil but works). The C# compiler split makes sure that that the pointer you get back skips the object header and length.

Alternately, there's likely (assuming, because I'm not that familiar with the Roslyn code base) an incremental way to read data that doesn't result in the whole thing requiring to be in-memory.